### ISSUE TITLE
`getStepTargetElement` for next step is called before `step:after` and `step:before` callbacks

### ISSUE THREAD

**Issue by srushingiv** [gilbarbara/react-joyride#241]
Since `joyride:getStepTargetElement` for next step is called before `step:after` and `step:before` callbacks ([in `componentWillUpdate`](/gilbarbara/react-joyride/blob/master/src/scripts/Joyride.jsx#L258)), designing tours in which components on the page change during the tour between steps results in errors thrown to the console stating "Target not mounted."

As a consumer of the Joyride API, I would like to be able to design a tour such that I can use the `step:after`, or even the `step:before` callback to trigger changes in the state of my page's components (such as menu or dialog open states) and seamlessly continue the tour.

In other words, the only time I would expect a "Target not mounted" error would be *after* all of the callbacks between the previous step and the one requiring the target.

Thank you.

**Comment by srushingiv** [comment]
After looking into this further, I'm realizing that regardless of when this function is called in the `componentWillUpdate` handler, it won't make a difference because all of where this handler is invoked in the react lifecycle.

In my opinion, `componentWillUpdate` is very likely not a good place for callbacks that might affect the react Virtual DOM to occur. Not only that, but referencing already-rendered elements by querySelectors is probably a poor choice given that we should very likely be querying the VDOM instead.

In order to solve this, it may be necessary to use React Refs instead of selectors for steps that require live updating during callbacks.

**Comment by gilbarbara** [comment]
Feel free to open a PR but a lot will change in V2 regarding the lifecycle.

**Comment by bluetidepro** [comment]
@srushingiv Did you ever figure out a good fix to this, running into the same problems.

@gilbarbara Is there an ETA on V2?

**Comment by ralphsmith80** [comment]
@bluetidepro I just hacked a solution together for this. It's not pretty. 

The biggest hiccup with the current joyride implementation is that clicking the next button does not invoke the public event `this.joyride.next`. Instead it uses a private event handler `handleClickTooltip`. Additionally `preventDefault` and `stopPropagation` is used. 

I bring this up because the obvious JS approach would be to wrap `this.joyride.next` or `Joyride.prototype.next` and do whatever you wanted before or after the function invocation. Because `this.joyride.next` is not being used internally and `handleClickTooltip` is private wrapping isn't possible.

Using a custom event listener and event delegation doesn't work because of the `stopPropagation` call. 

## Solution
What you can do is add a custom event listener directly on the next and back buttons. Although you need to manage that because the back button is removed if you navigate back to the first step so you need to rebind it and clean up the old handler if you're being neat. 

For each step where it was relevant I added an `onNextClick` and `onBackClick` handler. The idea being each step knows how to handle it's own custom next/back logic. In my case it's a `history.push` or `history.goBack` call. 

Note I wrapped the next/back handlers in a Promise to give is a thenable interface. I did this because I didn't want to use `setTimeout(() => {}, 0)` to give my route time to render. Also it allowed be to invoke `this.joyride.next` after my route was invoked. This was done so I could have the same behavior when clicking on a hole as when clicking the next button.

```
import React from 'react'
import Joyride from 'react-joyride'
import { withTheme } from 'material-ui/styles'
import { withRouter } from 'react-router'
import 'react-joyride/lib/react-joyride-compiled.css'

const nextButtonSelector =
  'button.joyride-tooltip__button.joyride-tooltip__button--primary'
const backButtonSelector =
  'button.joyride-tooltip__button.joyride-tooltip__button--secondary'

class OnBoard extends React.Component {
  _steps = () => {
    const { month, theme, history } = this.props
    const style = {
      mainColor: theme.palette.primary.main,
      beacon: {
        inner: theme.palette.secondary.main,
        outer: theme.palette.secondary.main
      }
    }
    return [
      [
        {
          title: 'Take the tour',
          text: 'Go on a tour',
          selector: '.home-page__button',
          style: style,
          onNextClick: () =>
            new Promise((resolve, reject) => {
              history.push('/second-page')
              resolve()
            })
        },
        {
          title: 'Second page',
          text: "You're on the second page of the tour",
          selector: '.second-page__button',
          style: style,
          onBackClick: () =>
            new Promise((resolve, reject) => {
              history.goBack()
              resolve()
            })
        },
    ]
  }
  _handleNextClick = evt => {
    const { steps } = this.joyride.props
    const step = steps[this.joyride.state.index]
    const onNextClick =
      step.onNextClick || (() => new Promise((resolve, reject) => resolve()))
    return onNextClick()
  }
  _handleBackClick = evt => {
    const { steps } = this.joyride.props
    const step = steps[this.joyride.state.index]
    const onBackClick = step.onBackClick || (() => {})
    if (this.backButton && this.joyride.state.index === 1) {
      this.backButton.removeEventListener('click', this._handleNextClick)
      this.backButton = undefined
    }
    return onBackClick()
  }
  render() {
    const { show } = this.props
    return (
      <Joyride
        ref={c => (this.joyride = c)}
        autoStart={true}
        run={show}
        steps={this._steps()}
        showStepsProgress={true}
        showSkipButton={true}
        type="continuous"
        // debug={true}
        callback={evt => {
          if (!this.nextButton) {
            setTimeout(() => {
              if (this.nextButton) return
              this.nextButton = document.querySelector(nextButtonSelector)
              if (!this.nextButton) return
              this.nextButton.addEventListener('click', this._handleNextClick)
            }, 0)
          }
          if (!this.backButton) {
            setTimeout(() => {
              if (this.backButton) return
              this.backButton = document.querySelector(backButtonSelector)
              if (!this.backButton) return
              this.backButton.addEventListener('click', this._handleBackClick)
            }, 0)
          }
          switch (evt.type) {
            case 'hole:click':
              this._handleNextClick().then(() => this.joyride.next())
              break
            default:
              break
          }
        }}
      />
    )
  }
}
export default withRouter(withTheme()(OnBoard))
```

**Comment by gilbarbara** [comment]
Please try V2 with `npm i react-joyride@next` and let me know it works for this case

**Comment by ralphsmith80** [comment]
@gilbarbara do you have a basic readme for V2? 

I fixed the steps to use `target` and removed the `import 'react-joyride/lib/react-joyride-compiled.css'` that was throwing and error. I wasn't seeing joyride show up which might just be a CSS issue but short of reading through repo to discern what the new api is I'm at a loss.

A simple example of whatever you're using to launch and run a tour for testing should be sufficient.

**Comment by gilbarbara** [comment]
@ralphsmith80 
Not yet, I'm finishing all the API changes this week and I'll start working on documentation...

I have a complete demo here that could help: https://codesandbox.io/s/2zpjporp4p

**Comment by ralphsmith80** [comment]
@bluetidepro  As an update my posted solution works but there's too many edge cases to handle with keyboard navigation and hole clicking and it's making me sad.

The only way I can get a consistent flow is to ensure the tour can't be stopped in the middle (skipping is okay). That requires `disableOverlay` which prevents getting a `hole:click` event so I think that invalidates my solution.

**Comment by ralphsmith80** [comment]
@gilbarbara I took a look at your V2 demo. The fact that you're using `handleClickNextButton` to update state and pass that as a prop to Joyride to advance state is so much more flexible :thumbsup:! 
I hope there's a similar handler for `back` and `skip`.  A user will need to make inverse route changes too.

Custom components look like a welcome addition too. My latest sadness came from not being able to override the `z-index` on the tool tip based on the step. I have a responsive site that utilizes slide outs on smaller screens and some steps I want over the slide out and some under. I could break my site theme definitions to make it work, but I'm not going to do that yet. 

I'll try out V2 again.

**Comment by gilbarbara** [comment]
@ralphsmith80 Do you want a step to be under your element? Can you explain or post an example in codesandbox?

Using V2 in controlled mode can give you control over timing and styling for a particular step so you could stop the tour while the new slide appears and turn it back on once it is in place.

**Comment by ralphsmith80** [comment]
@gilbarbara My issues were referring to V1. From what I've seen of V2 a `next` and `back` handler along with custom components should solve the issues I've had so far. I wanted to wrap up my V1 work so I had something presentable before I delve into V2. Although going straight to V2 might have been a shorter path.

**Comment by gilbarbara** [comment]
Yeah, I think so. V1 is buggy and I won't fix anything ever again. 😄 

It took me months to put V2 together trying to fix all the positioning and state bugs and now I'm getting close to release it. You can track the progress in #291 and I'm working on the documentation right now.

**Comment by ralphsmith80** [comment]
@gilbarbara, V2 is way easier to use and the thenable interface on individual steps I posted above works like a charm with being able to control when `stepIndex` is updated. 

Can you give me a heads up on how to override the step style? Are you required to use custom components in that case? 

I see [default styles](https://github.com/gilbarbara/react-joyride/blob/V2/src/styles.js) but my attempt at just overriding the styles didn't work. One thing I'm not sure about is the use of [`style.options`](https://github.com/gilbarbara/react-joyride/blob/V2/src/styles.js#L38). Should that just be `styles`?

Here's a style object I'm passing to each step and to the Joyride component:

```
     mainColor: theme.palette.primary.main,
      overlay: { zIndex: 1300 },
      tooltip: { zIndex: 1300 },
      hole: { zIndex: 1300 },
      tooltipContainer: { zIndex: 1300 },
      beacon: {
        zIndex: 1300,
        inner: theme.palette.secondary.main,
        outer: theme.palette.secondary.main
      }
```

The `overlay` and `hole` styles take affect when passed to the Joyride component but no styling is being applied to the steps.

**Comment by gilbarbara** [comment]
@ralphsmith80 
First, forget about V1 syntax. It's completely different...

If you want to override all steps' styling, set a `styles` prop in the Joyride component.
```jsx
<Joyride
  styles={
    options: { // if you want to override shared options
	  color: '#000', // this will change the overall color for all individual elements.
	  zIndex: 200, // but don't change this unless you really need it since Joyride uses React Portals now.
	},
	beaconInner: { // however if you the beacons to have a different color than the rest, use these.
	  backgroundColor: '#0f4'
	},
	beaconOuter: {
	  backgroundColor: '#0f4'
	}
  }
...
/>
```

Or you can set a `styles` prop for each step that will override the global options.

**Comment by ralphsmith80** [comment]
@gilbarbara 

From your example I see I was misunderstanding the syntax. I was putting thinks at the top level of the `styles` object rather than nested in `styles.options`.

I gave that a shot and here's what I found (this is all on the `Joyride` component):

The color updates with:
```
styles={{
  options: {
    color: theme.palette.primary.main
  }
}
```

However the `beaconInner`/`beaconOuter` style doesn't apply unless you move it outside of the nested `options` object.

```
styles={{
  options: {
    // this doesn't work
    beaconOuter: {
      backgroundColor: theme.palette.secondary.main
    }
  }
  // This works
  beaconOuter: {
    backgroundColor: theme.palette.secondary.main
  }
}
```

Also specifically for the beacon I noticed setting the `backgroundColor` on the `beaconOuter` property doesn't really do what you would want. The `borderColor` is what matter for the `beaconOuter` prop, but the issue with that is that the `backgroundColor` is not being translated from that. Meaning that the translucent middle section still has the same color as the top level `color` prop.

I tried updating your test project https://codesandbox.io/s/2zpjporp4p unsuccessfully. 

As for the need for overriding the `zIndex`. I'm using the material-ui component library which is quite popular and the default theme has many things defined with a `zIndex` > 1000. 
https://material-ui-next.com/customization/default-theme/

In my tests setting the `zIndex` was ineffective. The overlay got the right `zIndex` as did the beacon button and the inner tooltip element, but the container elements did not update which result in beacons and tooltips being under slide out menus. 

I'm going to take a look at the custom components to see if that will help.

**Comment by gilbarbara** [comment]
@ralphsmith80 which version?

**Comment by gilbarbara** [comment]
Also I've added the documentation for V2 in the README after I released 2.0.0-7

**Comment by ralphsmith80** [comment]
@gilbarbara 2.0.0-6

I'll give 2.0.0-7 a try.

**Comment by ralphsmith80** [comment]
@gilbarbara is the latest v2 code published to npm?

I'm not running from the source and I still have the same issues.  Just as a quick check trying to set the color with `primaryColor` as per the docs isn't working.

e.g.
not working
```
styles={{
  options: {
    primaryColor: theme.palette.primary.main
  }
}}
```

working
```
styles={{
  options: {
    color: theme.palette.primary.main
  }
}}
```
